plugins/radare2/radare2_main.ml

```diff
@@ -0,0 +1,90 @@
+open Bap_core_theory
+open Core_kernel
+open Bap_future.Std
+open Bap.Std
+include Self()
+
+open KB.Syntax
+
+let agent =
+  KB.Agent.register ~package:"bap.std" "radare2-symbolizer"
+    ~desc:"extracts symbols radare2"
+
+let provide_roots funcs =
+  let promise_property slot =
+    KB.promise slot @@ fun label ->
+    KB.collect Theory.Label.addr label >>| function
+    | None -> None
+    | Some addr ->
+      let addr = Bitvec.to_bigint addr in
+      Option.some_if (Hashtbl.mem funcs addr) true in
+  promise_property Theory.Label.is_valid;
+  promise_property Theory.Label.is_subroutine
+
+exception Radare2_failed
+
+let provide_radare2 file = 
+  let funcs = Hashtbl.create (module struct
+      type t = Z.t
+      let compare = Z.compare and hash = Z.hash
+      let sexp_of_t x = Sexp.Atom (Z.to_string x)
+    end) in
+    let accept name addr = Hashtbl.set funcs addr name in
+    let extract name json = let open Yojson in
+      match json with
+      | `Assoc list -> 
+        (match List.find list ~f:(fun (key, _) -> String.equal key name) with
+        | Some (_, v) -> Some v
+        | _ -> None)
+      | _ -> None in
+    try
+     let symbol_list = let open Yojson in
+      match R2.with_command_j "isj" file with
+      | `List list -> Some list
+      | _ -> None in
+     let strip str = let open String in
```

Comment:
Your helper function such as `strip`, `to_string`, et alas, are not really using the enclosed scope (i.e., the do not depend on the file), so it is better to move them away and define them on the toplevel. This will make the `provide_radare2` function easier to read and maintain, since now it will do only the work that depends on the file.